src/main/java/com/google/sps/data/TimeRange.java

```diff
@@ -0,0 +1,143 @@
+
+import java.time.Instant;
+import java.time.Duration;
+import java.util.Comparator;
+
+/**
+ * Class representing a span of time, enforcing properties (e.g. start comes before end) and
+ * providing methods to make ranges easier to work with (e.g. {@code overlaps}).
+ */
+public final class TimeRange {
+
+  /** A comparator for sorting ranges by their start time in ascending order. */
+  public static final Comparator<TimeRange> ORDER_BY_START =
+      new Comparator<TimeRange>() {
+        @Override
+        public int compare(TimeRange a, TimeRange b) {
+          return a.start.compareTo(b.start);
+        }
+      };
+
+  /** A comparator for sorting ranges by their end time in ascending order. */
+  public static final Comparator<TimeRange> ORDER_BY_END =
+      new Comparator<TimeRange>() {
+        @Override
+        public int compare(TimeRange a, TimeRange b) {
+          return a.end().compareTo(b.end());
+        }
+      };
+
+  private final Instant start, end;
+
+  /** Creates a timerange with a start and end instant */
+  private TimeRange(Instant start, Instant end) {
+    this.start = start;
+    this.end = end;
+  }
+
+  /** Returns the start of the range as an instant. */
+  public Instant start() {
+    return this.start;
+  }
+
+  /** Returns the number of minutes between the start and end of the range. */
+  public Long duration() {
+    return (Long) Duration.between(start, end).toMinutes();
+  }
+
+  /** Returns the end of the range as an instant. */
+  public Instant end() {
+    return this.end;
+  }
+
+  /**
+   * Checks if two ranges overlap. This means that at least some part of one range falls within the
+   * bounds of another range.
+   */
+  public boolean overlaps(TimeRange other) {
+    // For two ranges to overlap, one range must contain the start of another range.
+    //
+    // Case 1: |---| |---|
+    //
+    // Case 2: |---|
+    //            |---|
+    //
+    // Case 3: |---------|
+    //            |---|
+    return this.contains(other.start) || other.contains(this.start);
+  }
+
+  /**
+   * Checks if this range completely contains another range. This means that {@code other} is a
+   * subset of this range. This is an inclusive bounds, meaning that if two ranges are the same,
+   * they contain each other.
+   */
+  public boolean contains(TimeRange other) {
+    // If this range has no duration, it is irrelevant.
+    if (Duration.between(range.start, range.end).toMinutes() == 0) {
+      return false;
+    }
+
+    // If the other range has no duration, then we must treat it like a point in time rather than a
+    // range.
+    if (Duration.between(other.start, other.end) == 0) {
+      return contains(this, other.start);
+    }
+
+    // We need the inclusive end for this check in order for this case to equal true:
+    // |------|
+    //     |--|
+    Instant otherInclusiveEnd = other.end;
+    return contains(this, other.start) && contains(this, otherInclusiveEnd);
```

Comment:
This line should be able to handle the two if's you do above. lf any range is empty, it won't be able to contain anything. l'm fairly confident but make some tests just to double check?

Reply:
Maybe add some tests so we can just try to drop those two guard clauses.

Reply:
This line should be able to catch every (most?) cases you define above. Please remove as much as you can for simpler code.